### ISSUE TITLE
Tutorial?

### ISSUE THREAD

**Issue by cport1** [iodide-project/pyodide#357]
Is there a simple tutorial somewhere that shows how to take a simple python file that utilizes pyodide, pandas and numpy, takes a simple input, and outputs it to the window?

**Comment by mdboom** [comment]
https://alpha.iodide.io/notebooks/300/

We also had the "Hipster band finder" demo that utilizes pandas and D3 together that needs to be revived and put in the latest versions of Iodide.

**Comment by mdboom** [comment]
See: https://alpha.iodide.io/notebooks/1623/

**Comment by PythonLinks** [comment]
I have looked at both of those tutorials.   They are very good steps in the right direction.

You might find this video very interesting.  He is an experienced Jupyter Notebooks developer who 
talks about the problem with Jupyter Notebooks Tutorials. 
[I don't like notebooks](https://www.youtube.com/watch?v=7jiPeIFXb6U).- Joel Grus (Allen Institute for Artificial Intelligence)

You run a bunch of cells and you are done.   Something a bit different is needed.  A series of small exercises that build student knowledge and confidence. 
Based on recent experience, I share his point of view. 
I just gave a class on Pug.js yesterday.  
https://pythonlinks.info/pug-and-iodide-classes
I am about to start giving classes on Iodide and PyOdide.

What the students want is a class notes document, and a series of exercises.  Maybe multiple exercise use the same sandbox.  Maybe  there are multiple sandboxes. My Pug.js class has 3 sandboxes.  One is an in-browser demo, , one for server side page rendering (Python) and , the other for client side page rendering (Javascript).  From that I can say that you are on the right track.  You have the Iodide Tutorial, and two Pyodide Tutorials. 

It is not clear to me exactly what the course materials should include.  I think basically your demos, but where the student has to make some changes in each cell.  In the class yesterday, I was literally deleting lines of code, for the students to put back in. 

The bottom line is that I think that your demos are brilliant, they are a good step in the right direction for a tutorial, but more could be done to create tutorials.

**Comment by PythonLinks** [comment]
I got two likes on my last posting.   That is progress! Thank you. 

I have been thinking more about this tutorial.  First of all who is the target audience?  Jupiter Notebook/ Python data scientists moving to Iodide, smart developers, but often with little or no Javascript  experience, and a little scared of Javascript.  I used to be scared of Javascript.  I still am! 

We need to tell them that Javascript functions are easy to use.  It is a dynamically bound language much like Python but with a lot more brackets.  Nothing to be scared of.  Not as nice an object model as Python, stay away from Prototypes, but functions are easy to use.  The debugger is great. 

What do we want to accomplish? Give them a series of exercises that make them comfortable with the platform.  Debugging is best in Javascritpt, so as far as reasonable the tutorial should be in Javascript,  but their knowledge is of Python libraries, so we should use those. 
Things that are not easy to do in javascript, things for which no good Javascript library exists we should also do in Python.  

And of course the Tutorial should use the mainstream Pandas, some functions which are not really available in Javascript.  Could someone with expertise in this area please recommend two or three pandas functionalities which meet that criteria.  Omiting rows, and operator overloading come to mind, but I am not an expert in this area. 

Here is my idea for the exercises.  

1. In Javascript, download some data, log it to the console.  Make sure it works. Maybe they do not know how to use the browser debugger/console.  Before the class ask them to read a basic page on using the Google Chrome debugging tools. 

2.  Run the debugger on the above simple program.  Maybe they have never used the javascript debugger before.  Just spend a few minutes on this to give them the idea.  The tutorial is mostly about PyOdide.

2. Display the data. I am not sure if we should use the Javascript Plotly or Python’s MatPlotLib.  The later they are familiar with.  The former may be easier.  My instinct is to go with MatPlotLib, that is what they know and want. Human Factors! But do most of the calling from Javascript, that has the debugger, not from Python.   The existing  PyOdide demos are pretty Python heavy.

Pass the data to Pandas.  Do some simple calculation that is fast, and not available in Javascript.  We do not want to waste time in class waiting on computations.  Most of the demos are too slow for a tutorial.  Particularly because students have to rerun things multiple times to get it right. 

In Javascript, get the results back from Pandas  and then call the graphing routine on the results. 

Again in Javascript, add a simple form with a text box, and a slider.  Enter a number in the text box and slide the slider, and watch the graph change. 

And if there is time show them how to modify the  Iodide menu structure by editing the Pug.js.  
https://demo.forestwiki.com/Products/Iodide/Scripts/navbar/aceedit

You can see the resulting menus here:
https://demo.forestwiki.com/Iodide-notebooks/the-lorenz-attractor

Best to first login first to see how rich these menus are. 
User: admin
Password: admin

Note that I called the existing “tutorials” demos.  They are great demos because they demonstrate many features of the software.  That is different from a tutorial , which a new user needs to master in a few hour class.  Tutorials need to be simple.  Demos need to be rich.  It would be good to rename the existing “tutorials” to demos. 

The tutorial requirements would be to bring a laptop, know how to use Pandas, and MatPlotLib.  No Javascript experience is necessary, but do read a page on the Chrome Debugging tools before coming to class.  I have to find the page to recommend. 

So which graphing library do you think it should use?
Which Panda functions do you think it should use?
Do you know a good simple data set to support this tutorial?  
What should they calculate?

Is anyone else anywhere else in the world planning on offering PyOdide tutorials?  We could work together to create the course notes.

Thanks again for the likes
Chris
